src/http_parser.rs

```diff
@@ -1,199 +1,215 @@
-                b':' => {
-                    name = Span(start, pos);
-                    // skip leading whitespace on value
-                    while let Some(&b' ') | Some(&b'\t') = buffer.get(pos + 1) {
-                        pos += 1;
-                    }
-                    parsing_key = false;
-                    start = pos + 1;
-                }
-                b'\r' | b'\n' | b' ' | b'\t' => return Err(Error::ParseHeaderName),
-                _ => {}
-            }
-        } else if *c == b'\n' || (*c == b'\r' && buffer.get(pos + 1) == Some(&b'\n')) {
-            if name.is_empty() {
-                return Err(Error::ParseError);
-            }
 
-            let value = Span(start, pos);
-            headers.push((name, value));
-            if name.from_buf(&buffer).to_ascii_lowercase() == "content-length" {
-                content_length = value.from_buf(&buffer).parse().unwrap_or(0);
-            }
+    loop {
+        let name = parse_header_name(&buffer, &mut pos)?;
+        consume!(":");
+        consume_whitespace_to_eol!();
+        let value = parse_header_value(&buffer, &mut pos)?;
 
-            name = Span::new();
-            parsing_key = true;
-
-            // skip \r\n or \n
-            pos += if *c == b'\n' { 1 } else { 2 };
-
-            if let Some(next) = buffer.get(pos) {
-                match *next {
-                    b'\n' => {
-                        pos += 1;
-                        saw_end = true;
-                        break;
-                    }
-                    b'\r' => {
-                        if buffer.get(pos + 1) != Some(&b'\n') {
-                            return Ok(Status::Partial(buffer));
-                        }
-                        pos += 2;
-                        saw_end = true;
-                        break;
-                    }
-                    _ => {}
-                }
-            }
+        if name.from_buf(&buffer).to_ascii_lowercase() == "content-length" {
+            content_length = value.from_buf(&buffer).parse().unwrap_or(0);
+        }
+        headers.push((name, value));
 
-            start = pos;
-            continue;
+        consume_eol!(return Err(Error::ConnectionClosed));
+
+        if consume_eol!() {
+            break;
         }
-        len += 1;
-        if len > MAX_HEADER_SIZE {
+        if pos - start > MAX_HEADER_SIZE {
             return Err(Error::ParseHeaderValue);
         }
-        pos += 1;
-    }
-
-    // didn't receive full headers, abort
-    if !saw_end {
-        return Ok(Status::Partial(buffer));
     }
 
-    let method = Span(0, method_len);
-    let path = Span(method_len + 1, method_len + 1 + path_len);
     let body = if content_length > 0 {
-        Span(pos, pos + buffer.len())
+        Span(pos, pos + content_length)
     } else {
         Span::new()
     };
 
-    Ok(Status::Complete(Request::new(
+    if body.1 > buffer.len() {
+        return Err(ConnectionClosed);
```

Comment:
What does this do? Sorry, long day :sweat: 

To me it seems like if the span's length is larger than the input u8 data is, it returns connection closed. At first I thought this would make it so open threads sending nothing get killed but this is too far into the parse path (the `Vec<u8>` already exists).

Reply:
In the original source this was done up at the request layer once the request object is returned. This handles requests that are cut off before they have been completely sent. So if we have all the headers and the headers state the body of the request is larger than what we have, we have to assume that the connection was closed on the client-side before we received the full request.